### PR TITLE
detector: merge with cml-005-drop-stable

### PR DESCRIPTION
**Pull request by mrajwa**
This PR merges the detector component to the state from KWD release (cml-005-drop-stable)

Signed-off-by: Marcin Rajwa <marcin.rajwa@linux.intel.com>

### CODE REVIEW

File: tools/topology/m4/detect_test_coef.m4 [hunk 1]
```diff
@@ -6,6 +6,6 @@ CONTROLBYTES_PRIV(DETECTOR_priv,
 `       0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,'
 `       0x20,0x00,0x00,0x00,0x00,0x00,0x00,0x00,'
 `       0x00,0x00,0x00,0x00,0x34,0x08,0x00,0x00,'
-`       0x03,0x00,0xff,0x3f,0x34,0x08,0x00,0x00,'
-`       0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00"'
-)
+`       0x03,0x00,0x18,0x00,0x00,0x00,0x00,0x00,'
+`	0x34,0x08,0x00,0x00,0x00,0x00,0x00,0x00"'
+)
```
Comment (tlauda) [inline tools/topology/m4/detect_test_coef.m4]: New line missing.

File: src/include/user/detect_test.h [hunk 2]
```diff
@@ -5,10 +5,8 @@
  * Author: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>
  */
 
-#ifndef __USER_DETECT_TEST_H__
-#define __USER_DETECT_TEST_H__
-
-#include <stdint.h>
+#ifndef __INCLUDE_UAPI_USER_DETECT_TEST_H__
```
Comment (tlauda) [inline src/include/user/detect_test.h]: I think the rebase has overwritten changes from master.

File: src/audio/detect_test.c [hunk 3]
```diff
@@ -37,11 +37,15 @@
 #define tracev_keyword(__e, ...) \
 	tracev_event(TRACE_CLASS_KEYWORD, __e, ##__VA_ARGS__)
 
+#define INT24_MAX (0xFFFFFF / 2)
```
Comment (tlauda) [inline src/audio/detect_test.c]: It's already defined in sof/audio/format.h.

File: src/audio/detect_test.c [hunk 4]
```diff
@@ -668,8 +668,21 @@ static int test_keyword_reset(struct comp_dev *dev)
...
 static int test_keyword_prepare(struct comp_dev *dev)
 {
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint16_t valid_bits = dev->params.sample_valid_bytes * 8;
+	uint16_t sample_width = cd->config.sample_width;
+
 	trace_keyword("test_keyword_prepare()");
 
+	if (valid_bits != sample_width) {
+		/* Default threshold values has to be changes
```
Comment (tlauda) [inline src/audio/detect_test.c]: Typos:
values -> value
changes -> changed

File: src/audio/detect_test.c [hunk 5]
```diff
@@ -76,6 +76,25 @@ struct comp_data {
 			    struct comp_buffer *source, uint32_t frames);
 };
 
+static inline bool detector_is_sample_width_supported(enum sof_ipc_frame sf)
+{
+	bool ret;
```
Comment (tlauda) [inline src/audio/detect_test.c]: #include <stdbool.h>

File: src/audio/detect_test.c [hunk 6]
```diff
@@ -76,6 +76,25 @@ struct comp_data {
...
+
+	switch (sf) {
+	case SOF_IPC_FRAME_S16_LE:
+	/* FALLTHRU */
```
Comment (tlauda) [inline src/audio/detect_test.c]: Also for `SOF_IPC_FRAME_S16_LE`?

File: src/include/user/detect_test.h (hunk 2 above)
Comment (mrajwa) [reply src/include/user/detect_test.h]: Yes, corrected.

File: src/audio/detect_test.c (hunk 3 above)
Comment (mrajwa) [reply src/audio/detect_test.c]: Ok, I will use that one them.

**Comment by mrajwa** [comment]
@tlauda, updated.

File: src/audio/detect_test.c (hunk 6 above)
Comment (jajanusz) [reply src/audio/detect_test.c]: It should be in cases f.e.:
```
case SOF_IPC_FRAME_S16_LE:
	/* FALLTHRU */
case SOF_IPC_FRAME_S24_4LE:
	/* FALLTHRU */
```

Comment (mrajwa) [reply src/audio/detect_test.c]: @jajanusz @tlauda updated.

File: src/audio/detect_test.c [hunk 7]
```diff
@@ -76,6 +77,26 @@ struct comp_data {
 			    struct comp_buffer *source, uint32_t frames);
 };
 
+static inline bool detector_is_sample_width_supported(enum sof_ipc_frame sf)
```
Comment (jajanusz) [inline src/audio/detect_test.c]: Just a note: when function is static/private its name does not have to be unique, so you can make it shorter

File: src/audio/detect_test.c [hunk 8]
```diff
@@ -142,26 +142,34 @@ static void default_detect_test(struct comp_dev *dev,
...
 	struct comp_data *cd = comp_get_drvdata(dev);
-
-	int16_t *src;
-	int16_t diff;
+	void *src;
+	int32_t diff;
 	uint32_t count = frames; /**< Assuming single channel */
 	uint32_t sample;
+	uint16_t valid_bits = dev->params.sample_valid_bytes * 8;
+	int32_t activation_threshold = cd->config.activation_threshold;
```
Comment (jajanusz) [inline src/audio/detect_test.c]: I see it's helper read-only val. I'd make it const if checkpatch allows.

File: src/audio/detect_test.c [hunk 9]
```diff
@@ -72,6 +77,26 @@ struct comp_data {
 			    struct comp_buffer *source, uint32_t frames);
 };
 
+static inline bool detector_is_sample_width_supported(enum sof_ipc_frame sf)
+{
+	bool ret;
+
+	switch (sf) {
+	case SOF_IPC_FRAME_S16_LE:
+	/* FALLTHRU */
```
Comment (jajanusz) [inline src/audio/detect_test.c]: indent pls

File: src/audio/detect_test.c [hunk 10]
```diff
@@ -28,6 +28,8 @@
 #include <stddef.h>
 #include <stdint.h>
 #include <stdlib.h>
+#include <sof/audio/format.h>
```
Comment (tlauda) [inline src/audio/detect_test.c]: Should be after sof/audio/component.h.

File: src/audio/detect_test.c [hunk 11]
```diff
@@ -28,6 +28,8 @@
...
+#include <stdbool.h>
```
Comment (tlauda) [inline src/audio/detect_test.c]: Should be after errno.h.

File: src/audio/detect_test.c [hunk 12]
```diff
@@ -659,8 +697,21 @@ static int test_keyword_reset(struct comp_dev *dev)
...
+	uint16_t valid_bits = dev->params.sample_valid_bytes * 8;
+	uint16_t sample_width = cd->config.sample_width;
+
 	trace_keyword("test_keyword_prepare()");
 
+	if (valid_bits != sample_width) {
+		/* Default threshold value has to be changed
+		 * according to host new format.
+		 */
+		cd->config.activation_threshold = (valid_bits > 16U) ?
```
Comment (tlauda) [inline src/audio/detect_test.c]: Here you are using 16U and earlier just 16. Lack of consistency.

File: src/audio/detect_test.c (hunk 7 above)
Comment (mrajwa) [reply src/audio/detect_test.c]: I would keep it as it is, long but plain, simple and understandable at first sight.

File: src/audio/detect_test.c (hunk 12 above)
Comment (mrajwa) [reply src/audio/detect_test.c]: Here it is correct, I will force the constant to be unsigned in that other place too. Thanks for spotting.

File: src/audio/detect_test.c (hunk 8 above)
Comment (mrajwa) [reply src/audio/detect_test.c]: Can be done.

**Comment by mrajwa** [comment]
@tlauda @jajanusz updated.

**Comment by jajanusz** [comment]
Can squeeze it into minor 9. ABI change is not needed by kernel, so it can be merged as soon as passes CI.

File: tools/topology/m4/detect_test_coef.m4 [hunk 13]
```diff
@@ -6,6 +6,6 @@ CONTROLBYTES_PRIV(DETECTOR_priv,
 `       0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,'
 `       0x20,0x00,0x00,0x00,0x00,0x00,0x00,0x00,'
 `       0x00,0x00,0x00,0x00,0x34,0x08,0x00,0x00,'
-`       0x03,0x00,0xff,0x3f,0x34,0x08,0x00,0x00,'
-`       0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00"'
+`       0x03,0x00,0x18,0x00,0x00,0x00,0x00,0x00,'
```
Comment (keyonjie) [inline tools/topology/m4/detect_test_coef.m4]: are you setting activation_threshold to be 0 here?

Comment (mrajwa) [reply tools/topology/m4/detect_test_coef.m4]: yes its zero and FW uses default values.

**Pull request merged**